### PR TITLE
ORC-1130:[C++] Suppress the present stream when the data stream has no null value

### PR DESCRIPTION
**Pull request by coderex2522**
### Why are the changes needed?
if the data stream has no null value, the present stream of this column is redundant.

### How was this patch tested?
Through the UT TestReadIntent.testSuppressPresentStream.

### CODE REVIEW

File: c++/src/io/OutputStream.cc [hunk 1]
```diff
@@ -96,6 +96,9 @@ namespace orc {
     dataBuffer->resize(0);
     return dataSize;
   }
+  void BufferedOutputStream::suppress() {
```
Comment (dongjoon-hyun) [inline c++/src/io/OutputStream.cc:99]: Could you add an empty line before this?

Comment (coderex2522) [reply c++/src/io/OutputStream.cc:100]: done.

File: c++/src/ColumnWriter.cc [hunk 2]
```diff
@@ -139,10 +140,22 @@ namespace orc {
                          uint64_t offset,
                          uint64_t numValues,
                          const char* incomingMask) {
-    notNullEncoder->add(batch.notNull.data() + offset, numValues, incomingMask);
+    const char* notNull = batch.notNull.data() + offset;
+    notNullEncoder->add(notNull, numValues, incomingMask);
+    for (uint64_t i = 0; i < numValues; ++i) {
+      if (!notNull[i]) {
+        hasNullValue = true;
```
Comment (wgtmac) [inline c++/src/ColumnWriter.cc:147]: I'd suggest checking hasNullValue in line 145 or 146 because hasNullValue may already be set by previous nulls.

Comment (wgtmac) [reply c++/src/ColumnWriter.cc:147]: In addition, we should check batch.hasNulls as a shortcut.

File: c++/src/ByteRLE.cc [hunk 3]
```diff
@@ -203,6 +205,17 @@ namespace orc {
     recorder->add(static_cast<uint64_t>(numLiterals));
   }
 
+  void ByteRleEncoderImpl::suppress() {
+    // written data can be just ignored because they are only flushed in memory
+    outputStream->suppress();
+
+    numLiterals = 0;
```
Comment (wgtmac) [inline c++/src/ByteRLE.cc:212]: Why not nullify buffer? I'd add a reset() function to clear the state and use it here and in the ByteRleEncoderImpl constructor.

Comment (coderex2522) [reply c++/src/ByteRLE.cc:209]: I don't find the reset function, so I add this function by the way.

File: c++/test/TestReader.cc [hunk 4]
```diff
@@ -672,4 +672,71 @@ namespace orc {
...
+      ReaderOptions readerOptions;
+      readerOptions.setMemoryPool(*pool);
+      std::unique_ptr<Reader> reader =
+        createReader(std::move(inStream), readerOptions);
+      EXPECT_EQ(rowCount, reader->getNumberOfRows());
+      std::unique_ptr<RowReader> rowReader =
+        reader->createRowReader(RowReaderOptions());
+      auto batch = rowReader->createRowBatch(2048);
+      EXPECT_TRUE(rowReader->next(*batch));
+      EXPECT_EQ(rowCount, batch->numElements);
```
[the remark this answers is not in the document]
Comment (coderex2522) [reply c++/test/TestReader.cc]: Sure, based on your suggestion, I added more coverage to this test.

File: c++/test/TestReader.cc [hunk 5]
```diff
@@ -672,4 +672,71 @@ namespace orc {
...
+      if (!stripeFooter.ParseFromZeroCopyStream(pbStream.get())) {
+        throw ParseError("Parse stripe footer from pb stream failed");
+      }
+      for (int i = 0; i < stripeFooter.streams_size(); ++i) {
+        const proto::Stream& stream = stripeFooter.streams(i);
+        if (stream.has_kind() && stream.kind() == proto::Stream_Kind_PRESENT) {
+          EXPECT_EQ(stream.column(), 1UL);
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (coderex2522) [reply c++/test/TestReader.cc]: done.

File: c++/test/TestReader.cc [hunk 6]
```diff
@@ -672,4 +672,71 @@ namespace orc {
     EXPECT_EQ(1, nestedUnionBatch.offsets.data()[1]);
   }
 
+  TEST(TestReadIntent, testSuppressPresentStream) {
```
[the remark this answers is not in the document]
Comment (coderex2522) [reply c++/test/TestReader.cc]: I think this feature is mainly a modification on the writer side. So I move the test into TestWriter.cc

File: c++/src/ColumnWriter.cc [hunk 7]
```diff
@@ -139,10 +140,22 @@ namespace orc {
                          uint64_t offset,
                          uint64_t numValues,
                          const char* incomingMask) {
-    notNullEncoder->add(batch.notNull.data() + offset, numValues, incomingMask);
+    const char* notNull = batch.notNull.data() + offset;
+    notNullEncoder->add(notNull, numValues, incomingMask);
+    for (uint64_t i = 0; i < numValues; ++i) {
```
[the remark this answers is not in the document]
Comment (coderex2522) [reply c++/src/ColumnWriter.cc]: Done.

File: c++/src/ColumnWriter.cc (hunk 2 above)
Comment (coderex2522) [reply c++/src/ColumnWriter.cc:148]: This logic has been changed to take the or operation with batch.hasNulls.

**Review by wgtmac — APPROVED** [review]
LGTM +1.

**Review by stiga-huang — APPROVED** [review]
LGTM. +1

**Comment by wgtmac** [comment]
I have merged it. Thanks @coderex2522 @dongjoon-hyun @stiga-huang !

**Comment by dongjoon-hyun** [comment]
Thank you, @coderex2522 and @wgtmac !

**Pull request merged**